gh-46236: Add missing PyUnicode_Append() doc (#130531)
python/cpython: Doc/c-api/unicode.rst

   decref'ing the returned objects.

.. c:function:: void PyUnicode_Append(PyObject **p_left, PyObject *right)

   Append the string *right* to the end of *p_left*.  (In particular, a :exc:`SystemError`

>    if *p_left* or *right* is ``NULL``, or if either is not a Unicode object).

Reviewer: A little less wordy (English is hard!), and clarify a few more things:

```suggestion
   Append the string *right* to the end of *p_left*.
   *p_left* must point to a :term:`strong reference` to a Unicode object.

   On error, set *p_left* to ``NULL`` and set an exception.
```

Reviewer 2: I've taken the liberty of reflowing your text Peter. The HTML will be almost identical but the rst will be more readable (other functions indicate what happens in case of an error in another paragraph).

Reviewer: I agree with that now in hindsight, thanks :)

   decref'ing the returned objects.

.. c:function:: void PyUnicode_Append(PyObject **p_left, PyObject *right)

   Append the string *right* to the end of *p_left*.
   *p_left* must point to a :term:`strong reference` to a Unicode object.

>    On error, set *p_left* to ``NULL`` and set an exception.

Reviewer: ```suggestion
   On error, set *p_left* to ``NULL`` (*stealing* the reference) and set an exception.
```

> .. c:function:: void PyUnicode_Append(PyObject **p_left, PyObject *right)
> 
>    Append the string *right* to the end of *p_left*.
>    *p_left* must point to a :term:`strong reference` to a Unicode object.
> 
>    On error, set *\*p_left* to ``NULL`` and set an exception.

Reviewer 3: IMO, this needs to say that the reference is released in the error case. Merely setting a pointer no NULL doesn't do that.

How about this:
```suggestion
.. c:function:: void PyUnicode_Append(PyObject **p_left, PyObject *right)
   Append the string *right* to the end of *p_left*.
   *p_left* must point to a :term:`strong reference` to a Unicode object;
   :c:func:`!PyUnicode_Append` releases ("steals") this reference.

   On error, set *\*p_left* to ``NULL`` and set an exception.

   On sucess, set *\*p_left* to a new strong reference to the result.
```
(There's an optimization where if the "stolen" and "new" refs are the same object, their refcounting operations can be skipped. But externally, the function behaves as above.)

>    On error, set *p_left* to ``NULL`` (*stealing* the reference) and set an exception.

Reviewer 4: I don't understand the "stealing the reference" part. It's stolen to be put where? The reference is just destroyed by Py_DECREF() (via Py_CLEAR()).

Nitpick: the assignment is not on "p_left" but "*p_left". You should write `set *\*p_left*`.

Reviewer: I suggested that part--I'd also be ok with "releasing the reference," but I'm worried that's less clear.

Author: I noticed that `Py_DECREF(left)` exists only in the `else` block, so in general, the reference count of `p_left` should remain unchanged. I removed the ambiguous borrowed reference to avoid any potential misunderstandings.
(I hope I haven't misunderstood the code.)

```
    if (unicode_modifiable(left)
        && PyUnicode_CheckExact(right)
        && PyUnicode_KIND(right) <= PyUnicode_KIND(left)
        /* Don't resize for ascii += latin1. Convert ascii to latin1 requires
           to change the structure size, but characters are stored just after
           the structure, and so it requires to move all characters which is
           not so different than duplicating the string. */
        && !(PyUnicode_IS_ASCII(left) && !PyUnicode_IS_ASCII(right)))
    {
        /* append inplace */
        if (unicode_resize(p_left, new_len) != 0)
            goto error;

        /* copy 'right' into the newly allocated area of 'left' */
        _PyUnicode_FastCopyCharacters(*p_left, left_len, right, 0, right_len);
    }
    else {
        maxchar = PyUnicode_MAX_CHAR_VALUE(left);
        maxchar2 = PyUnicode_MAX_CHAR_VALUE(right);
        maxchar = Py_MAX(maxchar, maxchar2);

        /* Concat the two Unicode strings */
        res = PyUnicode_New(new_len, maxchar);
        if (res == NULL)
            goto error;
        _PyUnicode_FastCopyCharacters(res, 0, left, 0, left_len);
        _PyUnicode_FastCopyCharacters(res, left_len, right, 0, right_len);
        Py_DECREF(left);
        *p_left = res;
    }
```

.. c:function:: void PyUnicode_AppendAndDel(PyObject **p_left, PyObject *right)

>    The function is identical to :c:func:`PyUnicode_Append`, with the only

Reviewer 4: ```suggestion
   The function is similar to :c:func:`PyUnicode_Append`, with the only
```
